src/gluonts/model/n_beats/_network.py

```diff
@@ -0,0 +1,531 @@
+# Copyright 2018 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License").
+# You may not use this file except in compliance with the License.
+# A copy of the License is located at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# or in the "license" file accompanying this file. This file is distributed
+# on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+# express or implied. See the License for the specific language governing
+# permissions and limitations under the License.
+
+# Standard library imports
+from typing import List
+
+# Third-party imports
+import mxnet as mx
+import numpy as np
+
+# First-party imports
+from gluonts.core.component import validated
+from gluonts.model.common import Tensor
+from gluonts.evaluation._base import get_seasonality
+
+VALID_N_BEATS_STACK_TYPES = "G", "S", "T"
+VALID_LOSS_FUNCTIONS = "sMAPE", "MASE", "MAPE"
+
+
+def linear_space(
+    F, backcast_length: int, forecast_length: int, fwd_looking: bool
+):
+    if fwd_looking:
+        return F.arange(0, forecast_length) / forecast_length
+    else:
+        return F.arange(-backcast_length, 0) / backcast_length  # Option 01
+        # return F.arange(backcast_length, 0, -1) / backcast_length # Option 02
+        # return F.arange(0, backcast_length) / backcast_length # Option 03
+
+
+def seasonality_model(
+    F,
+    thetas: Tensor,
+    num_coefficients: int,
+    context_length: int,
+    prediction_length: int,
+    is_forecast: bool,
+) -> Tensor:
+    """
+    Creates a fourier series basis with num_coefficients coefficients for sine and cosine each.
+    So the total number of learned coefficients amounts to 2*num_coefficients.
+    """
+    t = linear_space(
+        F, context_length, prediction_length, fwd_looking=is_forecast
+    )
+    cosines = F.stack(
+        *[F.cos(2 * np.pi * i * t) for i in range(num_coefficients)]
+    )
+    sines = F.stack(
+        *[F.sin(2 * np.pi * i * t) for i in range(num_coefficients)]
+    )
+    S = F.concat(cosines, sines, dim=0)
```

Comment:
All these "feature" tensors are allocated every time the `NBEATSBlock` is invoked, so this could be optimized (same for `trend_model`). In the end, in interpretable modes, computing the forecast and backcast amounts to doing the final `F.dot` operation between the `thetas` and the block-type-specific feature tensor (lines 63 and 81 here) which can be pre-allocated. In the generic mode this is instead a dense layer.

One option is the following:
* Define the `NBEATSBlock` base class, implementing the `hybrid_forward` pretty much as it is now.
* Extend the base class into `NBEATSGenericBlock`, `NBEATSTrendBlock`, `NBEATSSeasonalBlock`, each one allocating *at construction time* the appropriate feature tensors given the hyperparameters (for the interpretable modes) or constructing the dense layer (for the generic mode), in addition to constructing the dense layers used to compute theta.
* Have the network decide what blocks to fill in the stacks with.

This comes maybe at the cost of a bit of additional boilerplate code (one constructor for each block class), but I think it's better than using the allocating hybrid lambda blocks wrapping a lambda wrapping a function.

Reply:
Very good catch! I agree this makes sense,  besides performance this should improve code readability, since the conceptual segmentation of the different block types is also present in the class structure then.